edocumentation-service/src/main/java/org/dcsa/edocumentation/service/unofficial/ManageShipmentService.java

```diff
@@ -87,6 +92,66 @@ private String generateCarrierBookingReference() {
     return new String(cbr);
   }
 
+  private String generateCommoditySubreference() {
+    char[] ref = new char[100];
+    for (int i = 0 ; i < ref.length ; i++) {
+      ref[i] = VALID_CBR_AUTOGEN_CHARS[random.nextInt(VALID_CBR_AUTOGEN_CHARS.length)];
+    }
+    return new String(ref);
+  }
+
+  private void assignCommoditySubreferences(
+    Booking booking,
+    List<@Valid List<@Valid @NotBlank @Size(max = 100) @Pattern(regexp = "^\\S(\\s+\\S+)*$") String>> subreferences
+  ) {
+    if (subreferences != null) {
+      long totalCount = subreferences.stream().mapToLong(Collection::size).sum();
+      long uniqueCount = subreferences.stream()
+        .flatMap(Collection::stream)
+        .distinct()
+        .count();
+      if (totalCount != uniqueCount) {
+        throw ConcreteRequestErrorMessageException.invalidInput("All commoditySubreferences must be unique"
+          + " across the shipment");
+      }
+      var requestedEquipments = booking.getRequestedEquipments();
+      if (subreferences.size() != requestedEquipments.size()) {
+        throw ConcreteRequestErrorMessageException.invalidInput("There are " + requestedEquipments.size()
+         + " requested equipment groups but subreferences for " + subreferences.size()
+         + " requested equipment groups. These two numbers should be the same.");
+      }
+
+      for (int i = 0 ; i < requestedEquipments.size() ; i++) {
+        var commodities = requestedEquipments.get(i).getCommodities();
+        var refs = subreferences.get(i);
+        if (commodities.size() != refs.size()) {
+          throw ConcreteRequestErrorMessageException.invalidInput("There are " + commodities.size()
+            + " commodities in but subreferences for " + refs.size()
+            + " commodities in the requested equipment group with index " + i + "."
+            + " These two numbers should be the same.");
+        }
+        var j = 0;
+        for (var commodity : commodities) {
+          var ref = refs.get(j);
+          if (commodity.getCommoditySubreference() != null && !commodity.getCommoditySubreference().equals(ref)) {
```

Comment:
can some ref variables in refs also be equal to null? If yes, please add check that corresponding commodity.getCommoditySubreference() is also null

Reply:
I do not understand this comment; There is literally a null check for the commodity subreference on the very same line...?

Reply:
I meant this use case:
if (commodity.getCommoditySubreference() == null && ref != null) {
// throw exception
}

Reply:
![billede](https://github.com/dcsaorg/DCSA-Edocumentation/assets/70640050/9af154d5-f426-4d2d-acd3-8b42cc693e7d)

`@NotBlank` implies `@NotNull`